src/approaches/grammar_search_invention_approach.py

```diff
@@ -432,93 +432,110 @@ def learn_from_offline_dataset(self, dataset: Dataset,
-            total_score = \
-                CFG.grammar_search_false_pos_weight * num_false_positives + \
-                CFG.grammar_search_true_pos_weight * (-num_true_positives) + \
-                CFG.grammar_search_size_weight * op_size + \
-                CFG.grammar_search_pred_complexity_weight * pred_complexity
-            # Useful for debugging:
-            # print("TP/FP/S/C/Total:", num_true_positives, num_false_positives,
-            #       op_size, pred_complexity, total_score)
-            # Lower is better.
-            return total_score
-
-        # There are no goal states for this search; run until exhausted.
-        def _check_goal(s: FrozenSet[Predicate]) -> bool:
-            del s  # unused
-            return False
-
-        if CFG.grammar_search_direction == "largetosmall":
-            # Successively consider smaller predicate sets.
-            def _get_successors(s: FrozenSet[Predicate]
-                    ) -> Iterator[Tuple[None, FrozenSet[Predicate], float]]:
-                for predicate in sorted(s):  # sorting for determinism
-                    # Actions not needed. Frozensets for hashing.
-                    # The cost of 1. is irrelevant because we're doing GBFS
-                    # and not A* (because we don't care about the path).
-                    yield (None, frozenset(s - {predicate}), 1.)
-
-            # Start the search with all of the candidates.
-            init = frozenset(candidates)
-        else:
-            assert CFG.grammar_search_direction == "smalltolarge"
-            # Successively consider larger predicate sets.
-            def _get_successors(s: FrozenSet[Predicate]
-                    ) -> Iterator[Tuple[None, FrozenSet[Predicate], float]]:
-                for predicate in sorted(set(candidates) - s):  # determinism
-                    # Actions not needed. Frozensets for hashing.
-                    # The cost of 1. is irrelevant because we're doing GBFS
-                    # and not A* (because we don't care about the path).
-                    yield (None, frozenset(s | {predicate}), 1.)
-
-            # Start the search with no candidates.
-            init = frozenset()
-
-        # Greedy best first search.
-        path, _ = utils.run_gbfs(
-            init, _check_goal, _get_successors, _heuristic,
-            max_evals=CFG.grammar_search_max_evals)
-        kept_predicates = path[-1]
-
-        print(f"Selected {len(kept_predicates)} predicates out of "
-              f"{len(candidates)} candidates:")
-        for pred in kept_predicates:
-            print(pred)
-
-        return set(kept_predicates)
+
+def _create_heuristic_function(initial_predicates: Set[Predicate],
+        atom_dataset: List[GroundAtomTrajectory],
+        candidates: Dict[Predicate, float]
+        ) -> Callable[[FrozenSet[Predicate]], float]:
+    if CFG.grammar_search_heuristic == "prediction_error":
+        return partial(_prediction_error_heuristic,
```

Comment:
clarify with a comment that this returns a function which takes in the last argument of _prediction_error_heuristic. It's not intuitive at a glance.